include/deal.II/sundials/n_vector.templates.h

```diff
@@ -0,0 +1,806 @@
+         *unwrap_nvector_const<VectorType>(y);
+}
+
+
+
+template <typename VectorType>
+void
+SUNDIALS::internal::NVectorOperations::set_constant(realtype c, N_Vector v)
+{
+  auto *v_dealii = unwrap_nvector<VectorType>(v);
+  *v_dealii      = c;
+}
+
+
+
+template <typename VectorType>
+void
+SUNDIALS::internal::NVectorOperations::add_constant(N_Vector x,
+                                                    realtype b,
+                                                    N_Vector z)
+{
+  auto *x_dealii = unwrap_nvector_const<VectorType>(x);
+  auto *z_dealii = unwrap_nvector<VectorType>(z);
+
+  if (x_dealii == z_dealii)
+    z_dealii->add(b);
+  else
+    {
+      *z_dealii = *x_dealii;
+      z_dealii->add(b);
+    }
+}
+
+
+
+template <typename VectorType>
+void
+SUNDIALS::internal::NVectorOperations::elementwise_product(N_Vector x,
+                                                           N_Vector y,
+                                                           N_Vector z)
+{
+  auto *x_dealii = unwrap_nvector_const<VectorType>(x);
+  auto *y_dealii = unwrap_nvector_const<VectorType>(y);
+  auto *z_dealii = unwrap_nvector<VectorType>(z);
+  if (z_dealii == x_dealii)
+    z_dealii->scale(*y_dealii);
+  else if (z_dealii == y_dealii)
+    z_dealii->scale(*x_dealii);
+  else
+    {
+      *z_dealii = *y_dealii;
+      z_dealii->scale(*x_dealii);
+    }
+}
+
+
+
+template <typename VectorType>
+realtype
+SUNDIALS::internal::NVectorOperations::weighted_rms_norm(N_Vector x, N_Vector w)
+{
+  // TODO copy can be avoided by a custom kernel
+  VectorType tmp      = *unwrap_nvector_const<VectorType>(x);
+  auto *     w_dealii = unwrap_nvector_const<VectorType>(w);
+  const auto n        = tmp.size();
+  tmp.scale(*w_dealii);
+  return tmp.l2_norm() / std::sqrt(n);
+}
+
+
+
+template <typename VectorType>
+realtype
+SUNDIALS::internal::NVectorOperations::max_norm(N_Vector x)
+{
+  return unwrap_nvector_const<VectorType>(x)->linfty_norm();
+}
+
+
+
+template <typename VectorType>
+void
+SUNDIALS::internal::NVectorOperations::scale(realtype c, N_Vector x, N_Vector z)
+{
+  auto *x_dealii = unwrap_nvector_const<VectorType>(x);
+  auto *z_dealii = unwrap_nvector<VectorType>(z);
+
+  if (x_dealii == z_dealii)
+    (*z_dealii) *= c;
+  else
+    z_dealii->sadd(0.0, c, *x_dealii);
+}
+
+
+
+template <typename VectorType,
+          std::enable_if_t<!IsBlockVector<VectorType>::value, int>>
+void
+SUNDIALS::internal::NVectorOperations::elementwise_div(N_Vector x,
+                                                       N_Vector y,
+                                                       N_Vector z)
+{
+  auto *x_dealii = unwrap_nvector_const<VectorType>(x);
+  auto *y_dealii = unwrap_nvector_const<VectorType>(y);
+  auto *z_dealii = unwrap_nvector<VectorType>(z);
+
+  AssertDimension(x_dealii->size(), z_dealii->size());
+  AssertDimension(x_dealii->size(), y_dealii->size());
+
+  std::transform(x_dealii->begin(),
```

Comment:
This is an example of a vector operation that is not easily translated to the dealii vector interface. I went for this naive implementation for now to have a working version in place.